wgpu/utils/shadertoy.py

```diff
@@ -0,0 +1,311 @@
+        )
+
+        shader_code = vertex_code + builtin_variables + self.shader_code + fragment_code
+        shader_program = self._device.create_shader_module(code=shader_code)
+
+        self._uniform_buffer = self._device.create_buffer(
+            size=self._uniform_data.nbytes,
+            usage=wgpu.BufferUsage.UNIFORM | wgpu.BufferUsage.COPY_DST,
+        )
+
+        bind_group_layout = self._device.create_bind_group_layout(
+            entries=binding_layout
+        )
+
+        self._bind_group = self._device.create_bind_group(
+            layout=bind_group_layout,
+            entries=[
+                {
+                    "binding": 0,
+                    "resource": {
+                        "buffer": self._uniform_buffer,
+                        "offset": 0,
+                        "size": self._uniform_data.nbytes,
+                    },
+                },
+            ],
+        )
+
+        self._render_pipeline = self._device.create_render_pipeline(
+            layout=self._device.create_pipeline_layout(
+                bind_group_layouts=[bind_group_layout]
+            ),
+            vertex={
+                "module": shader_program,
+                "entry_point": "vs_main",
+                "buffers": [],
+            },
+            primitive={
+                "topology": wgpu.PrimitiveTopology.triangle_list,
+                "front_face": wgpu.FrontFace.ccw,
+                "cull_mode": wgpu.CullMode.none,
+            },
+            depth_stencil=None,
+            multisample=None,
+            fragment={
+                "module": shader_program,
+                "entry_point": "fs_main",
+                "targets": [
+                    {
+                        "format": wgpu.TextureFormat.bgra8unorm,
+                        "blend": {
+                            "color": (
+                                wgpu.BlendFactor.one,
+                                wgpu.BlendFactor.zero,
+                                wgpu.BlendOperation.add,
+                            ),
+                            "alpha": (
+                                wgpu.BlendFactor.one,
+                                wgpu.BlendFactor.zero,
+                                wgpu.BlendOperation.add,
+                            ),
+                        },
+                    },
+                ],
+            },
+        )
+
+    def _bind_events(self):
+        def on_resize(event):
+            w, h = event["width"], event["height"]
+            self._uniform_data["resolution"] = (w, h)
+
+        def on_mouse_move(event):
+            xy = event["x"], event["y"]
+            if event["button"] == 1 or 1 in event["buttons"]:
+                self._uniform_data["mouse"] = xy
+
+        self._canvas.add_event_handler(on_resize, "resize")
+        self._canvas.add_event_handler(on_mouse_move, "pointer_move", "pointer_down")
+
+    def _update(self):
+        now = time.time()
```

Comment:
Only relative time matters, so let's use a more precise variant:
```suggestion
        now = time.perf_counter()
```